triedb/pathdb/history_indexer.go

```diff
@@ -59,91 +77,94 @@ func loadIndexMetadata(db ethdb.KeyValueReader) *indexMetadata {
+	delete  bool                    // Operation mode: true for unindex, false for index.
+	lastID  uint64                  // ID of the most recently processed history.
+	typ     historyType             // Type of history being processed (e.g., state or trienode).
+	db      ethdb.KeyValueStore     // Key-value database used to store or delete index data.
 }
 
 // newBatchIndexer constructs the batch indexer with the supplied mode.
-func newBatchIndexer(db ethdb.KeyValueStore, delete bool) *batchIndexer {
+func newBatchIndexer(db ethdb.KeyValueStore, delete bool, typ historyType) *batchIndexer {
 	return &batchIndexer{
-		accounts: make(map[common.Hash][]uint64),
-		storages: make(map[common.Hash]map[common.Hash][]uint64),
-		delete:   delete,
-		db:       db,
+		index:  make(map[stateIdent][]uint64),
+		delete: delete,
+		typ:    typ,
+		db:     db,
 	}
 }
 
-// process iterates through the accounts and their associated storage slots in the
-// state history, tracking the mapping between state and history IDs.
-func (b *batchIndexer) process(h *stateHistory, historyID uint64) error {
-	for _, address := range h.accountList {
-		addrHash := crypto.Keccak256Hash(address.Bytes())
-		b.counter += 1
-		b.accounts[addrHash] = append(b.accounts[addrHash], historyID)
-
-		for _, slotKey := range h.storageList[address] {
-			b.counter += 1
-			if _, ok := b.storages[addrHash]; !ok {
-				b.storages[addrHash] = make(map[common.Hash][]uint64)
-			}
-			// The hash of the storage slot key is used as the identifier because the
-			// legacy history does not include the raw storage key, therefore, the
-			// conversion from storage key to hash is necessary for non-v0 histories.
-			slotHash := slotKey
-			if h.meta.version != stateHistoryV0 {
-				slotHash = crypto.Keccak256Hash(slotKey.Bytes())
-			}
-			b.storages[addrHash][slotHash] = append(b.storages[addrHash][slotHash], historyID)
-		}
+// process traverses the state entries within the provided history and tracks the mutation
+// records for them.
+func (b *batchIndexer) process(h history, id uint64) error {
+	for ident := range h.forEach() {
+		b.index[ident] = append(b.index[ident], id)
+		b.pending++
 	}
-	b.lastID = historyID
+	b.lastID = id
+
 	return b.finish(false)
 }
 
 // finish writes the accumulated state indexes into the disk if either the
 // memory limitation is reached or it's requested forcibly.
 func (b *batchIndexer) finish(force bool) error {
-	if b.counter == 0 {
+	if b.pending == 0 {
 		return nil
 	}
-	if !force && b.counter < historyIndexBatch {
+	if !force && b.pending < historyIndexBatch {
 		return nil
 	}
 	var (
-		batch    = b.db.NewBatch()
-		batchMu  sync.RWMutex
-		storages int
-		start    = time.Now()
-		eg       errgroup.Group
+		batch   = b.db.NewBatch()
+		batchMu sync.RWMutex
+		start   = time.Now()
+		eg      errgroup.Group
 	)
 	eg.SetLimit(runtime.NumCPU())
 
-	for addrHash, idList := range b.accounts {
```

Comment:
here we slightly change the order, right? Previously we deleted all accounts then all storages, now we delete the first account and all their storage and then the second account. 

Ah wait it all happened in parallel with the errgroup anyway